echopype/echodata/zarr_combine.py

```diff
@@ -0,0 +1,742 @@
+        """
+
+        self._check_ascending_ds_times(ds_list, ed_name)
+        self._check_channels(ds_list, ed_name)
+
+        # Dataframe with column as dim names and rows as the different Datasets
+        self.dims_df = pd.DataFrame([ds.dims for ds in ds_list])
+
+        # calculate useful information about the dimensions
+        self.dims_sum = self.dims_df.sum(axis=0).to_dict()
+        self.dims_csum = self.dims_df.cumsum(axis=0).to_dict()
+        self.dims_max = self.dims_df.max(axis=0).to_dict()
+
+        # format ed_name appropriately
+        ed_name = ed_name.replace("-", "_").replace("/", "_").lower()
+
+        if len(ds_list) == 1:
+            # get numpy keys if we only have one Dataset
+            numpy_keys = self._compare_attrs(ds_list[0].attrs, ds_list[0].attrs)
+        else:
+            # compare attributes and get numpy keys, if they exist
+            for ind in range(len(ds_list) - 1):
+                numpy_keys = self._compare_attrs(ds_list[ind].attrs, ds_list[ind + 1].attrs)
+
+        # collect Dataset attributes
+        for count, ds in enumerate(ds_list):
+
+            # get reduced attributes that do not include numpy keys
+            red_attrs = {key: val for key, val in ds.attrs.items() if key not in numpy_keys}
+
+            if count == 0:
+                self.group_attrs[ed_name + "_attr_key"].extend(red_attrs.keys())
+            self.group_attrs[ed_name + "_attrs"].append(list(red_attrs.values()))
+
+    def _get_temp_arr(self, dims: List[str], dtype: type) -> Tuple[type(dask.array), list]:
+        """
+        Constructs a temporary (or dummy) array representing a
+        variable in its final combined form.
+
+        Parameters
+        ----------
+        dims: List[str]
+            A list of the dimension names
+        dtype: type
+            The data type of the variable
+
+        Returns
+        -------
+        temp_arr: dask.array
+            a temporary (or dummy) array representing a
+            variable in its final combined form.
+        chnk_shape: List[int]
+            The chunk shape used to construct ``temp_arr``
+
+        Notes
+        -----
+        This array is never interacted with in a traditional sense.
+        Its sole purpose is to construct metadata for the zarr store.
+        """
+
+        # Create the shape of the variable in its final combined form
+        shape = [
+            self.dims_sum[dim] if dim in self.append_dims else self.dims_max[dim] for dim in dims
+        ]
+
+        # Create the chunk shape of the variable
+        # TODO: investigate if this is the best chunking
+        chnk_shape = [self.dims_max[dim] for dim in dims]
+
+        temp_arr = dask.array.zeros(shape=shape, dtype=dtype, chunks=chnk_shape)
+
+        return temp_arr, chnk_shape
+
+    def _set_encodings(
+        self, encodings: Dict[str, dict], name: Hashable, val: xr.Variable, chnk_shape: list
+    ) -> None:
```

Comment:
I'm trying to wrap my head around this and `utils/io.py::set_zarr_encodings` that only gets used in `utils/io.py::save_file`, which is called when groups are saved to file using `EchoData.to_zarr/to_netcdf`. It seems the operations are partially overlapping so encodings for some variables are overwritten (maybe with the same settings?) when the data is actually written to disk using the `to_zarr/to_netcdf` functions?

Reply:
In this function I am gathering the encodings (except the lazy encodings) of each variable so that I can pass them off to the delayed `to_zarr` function. By supplying the encodings, we can construct all of the metadata for the final combined EchoData object. Here we do not want to use the encodings such as `chunks` because this encoding would correspond to the encoding of the first dataset in the datasets that are to be combined. Instead, we want to set chunks to be uniform (this is done at the end of the function).

> It seems the operations are partially overlapping so encodings for some variables are overwritten (with the same settings?)

We are not overwriting any encodings, just collecting them. The statement that sets the compressor is only needed if no compressor is provided. No compressor is present if the provided variable was obtained from an in-memory EchoData group i.e. we are combining in-memory EchoData objects.

Reply:
Ah now I see that the collection only happens here. Just to clarify: by "lazy encodings" do you mean like `chunk`?

Reply:
Also, after looking at the use of this function some more, I also realized why I was confused when first reading it. I think I was not expecting `encodings` in `_construct_lazy_ds_and_var_info` in modified in place with `self._set_encodings(encodings, name, val, chnk_shape)`. 

How about making this function to return a new dict (let's call it `new_dict` for now), and use `encodings.update(new_dict)` to add the new encoding key-value pair to the dict `encodings` in `_construct_lazy_ds_and_var_info`?

Reply:
> Ah now I see that the collection only happens here. Just to clarify: by "lazy encodings" do you mean like chunk?

Here, "lazy encodings" are `chunks` and `preferred_chunks`. 

I will address the `set_encodings` comment in the other more in-depth comment you make below.

Reply:
Thanks for the explanation!